### PR TITLE
Houston code

### PR DESCRIPTION
**Pull request by RobinsonZ**

### CODE REVIEW

File: src/main/java/org/team1540/robot2019/Hardware.java [hunk 1]
```diff
@@ -206,6 +206,22 @@ public static void initElevator() {
         logger.info("Initialized elevator in " + (end - start) + " ms");
     }
 
+    public static void unfollowInitElevator() {
```
Comment (RobinsonZ) [inline src/main/java/org/team1540/robot2019/Hardware.java]: Should probably be refactored to a verb phrase

File: src/main/java/org/team1540/robot2019/Hardware.java [hunk 2]
```diff
@@ -206,6 +206,22 @@ public static void initElevator() {
...
+        logger.info("Initializing elevator...");
```
Comment (RobinsonZ) [inline src/main/java/org/team1540/robot2019/Hardware.java]: Clarify these logs

File: src/main/java/org/team1540/robot2019/Hardware.java [hunk 3]
```diff
@@ -229,17 +245,21 @@ public static void initCargoMech() {
         logger.info("Initializing cargo mech...");
         double start = RobotController.getFPGATime() / 1000.0; // getFPGATime returns microseconds
 
-        cargoRollerTop = createController(RobotMap.CARGO_ROLLER_TOP);
+        //        cargoRollerTopTalon.configPeakCurrentLimit(50);
```
Comment (RobinsonZ) [inline src/main/java/org/team1540/robot2019/Hardware.java]:
```suggestion
```

File: src/main/java/org/team1540/robot2019/Hardware.java [hunk 4]
```diff
@@ -229,17 +245,21 @@ public static void initCargoMech() {
...
+
+        ChickenTalon cargoRollerTop = new ChickenTalon(RobotMap.CARGO_ROLLER_TOP);
+        Hardware.cargoRollerTop = cargoRollerTop;
```
Comment (RobinsonZ) [inline src/main/java/org/team1540/robot2019/Hardware.java]: Why qualified access?

Also should maybe check if that actually is a talon on the competition robot as I don't quite remember

File: src/main/java/org/team1540/robot2019/Hardware.java [hunk 5]
```diff
@@ -229,17 +245,21 @@ public static void initCargoMech() {
...
+        cargoRollerTop.configPeakCurrentLimit(50);
         cargoRollerBottom = createController(RobotMap.CARGO_ROLLER_BOTTOM);
 
-        cargoRollerTop.setInverted(Tuning.invertCargoRollerTop);
+        Hardware.cargoRollerTop.setInverted(Tuning.invertCargoRollerTop);
```
Comment (RobinsonZ) [inline src/main/java/org/team1540/robot2019/Hardware.java]: Same here with fully qualified access

File: src/main/java/org/team1540/robot2019/OI.java [hunk 6]
```diff
@@ -54,15 +55,16 @@
     // ---------------------------------------- Copilot ----------------------------------------
     // Hatch
     private static Button sensorGrabHatchButton = copilot.getButton(XboxButton.X);
-    private static Button prepGetHatchFloorButton = copilot.getButton(XboxButton.START);
+    private static Button placeHatchSequenceButton = copilot.getButton(XboxButton.START);
```
Comment (RobinsonZ) [inline src/main/java/org/team1540/robot2019/OI.java]: Rename, maybe `manualPlaceHatchButton` to contrast with `placeHatchButton`?

File: src/main/java/org/team1540/robot2019/OI.java [hunk 7]
```diff
@@ -54,15 +55,16 @@
...
     private static Button grabThenRetractButtonAndAlsoTheRocketBallIntakeButton = copilot.getButton(XboxAxis.RIGHT_TRIG, Tuning.axisButtonThreshold);
```
Comment (RobinsonZ) [inline src/main/java/org/team1540/robot2019/OI.java]: Make this name less horrible

File: src/main/java/org/team1540/robot2019/OI.java [hunk 8]
```diff
@@ -155,25 +157,30 @@ static void init() {
...
+        grabThenRetractButtonAndAlsoTheRocketBallIntakeButton.whenPressed(new GrabHatchThenRetract(Tuning.hatchGrabWaitTime));
+        stowHatchButton.whenPressed(new CommandGroup() {{
+            addSequential(new ReleaseHatch());
+            addSequential(new RetractHatchMech());
+        }});
+
+        //    Vision hatch place
+        VisionAutoPlaceSequence visionAutoPlaceSequence = new VisionAutoPlaceSequence();
+        visionPlaceHatchButton.whenPressed(visionAutoPlaceSequence);
+        visionPlaceHatchButton.whenReleased(new SimpleCommand("", visionAutoPlaceSequence::cancel));
```
Comment (RobinsonZ) [inline src/main/java/org/team1540/robot2019/OI.java]: This can happen with a `whileHeld()`

File: src/main/java/org/team1540/robot2019/OI.java [hunk 9]
```diff
@@ -186,6 +193,8 @@ static void init() {
         elevatorCargoShipButton.whenPressed(moveElevatorToCargoShip);
         MoveElevatorToZero moveElevatorToZero = new MoveElevatorToZero();
         elevatorDownButton.whenPressed(moveElevatorToZero);
+//        MoveElevatorToPosition moveElevatorToClimbArmsBack = new MoveElevatorToPosition(Tuning.elevatorToClimbArmsBack);
+//        moveElevatorToClimbArmsBackButton.whenPressed(moveElevatorToClimbArmsBack);
```
Comment (RobinsonZ) [inline src/main/java/org/team1540/robot2019/OI.java]: Is the moveElevator thing used? If not, should probably be deleted

File: src/main/java/org/team1540/robot2019/OI.java [hunk 10]
```diff
@@ -266,9 +276,11 @@ static void init() {
...
             testPrepGetHatchButton.whenPressed(sensorGrabHatchSequence);
             testPlaceHatchButton.whenPressed(placeHatchSequence);
-            testPlaceHatchInLoadingStationButton.whenPressed(new PlaceHatchInLoadingStation());
-//            autoPlaceButton.whenPressed(new VisionPlaceSequence());
-//            autoGrabButton.whenPressed(new VisionGrabSequence());
+            testPlaceHatchInLoadingStationButton.whenPressed(new VisionAutoPlaceSequence());
+//            testPlaceHatchInLoadingStationButton.whenPressed(new PlaceHatchInLoadingStation());
+
+//            autoPlaceButton.whenPressed(new VisionAutoPlaceSequence());
+//            autoGrabButton.whenPressed(new VisionAutoGrabSequence());
```
Comment (RobinsonZ) [inline src/main/java/org/team1540/robot2019/OI.java]: Remove commented code

File: src/main/java/org/team1540/robot2019/Robot.java [hunk 11]
```diff
@@ -238,17 +242,14 @@ public void teleopPeriodic() {
     private void compressorPeriodic() {
         if (SmartDashboard.getBoolean("EnableCompressor", true)) {
```
Comment (RobinsonZ) [inline src/main/java/org/team1540/robot2019/Robot.java]: This `if` could be condensed

File: src/main/java/org/team1540/robot2019/Tuning.java [hunk 12]
```diff
@@ -52,6 +52,7 @@
 
     public static double elevatorUpPosition = 28;
     public static double elevatorCargoShipPosition = 16;
+    public static double elevatorToClimbArmsBack = 10;
```
Comment (RobinsonZ) [inline src/main/java/org/team1540/robot2019/Tuning.java]: Should be `elevatorClimbArmsBackPosition`

File: src/main/java/org/team1540/robot2019/commands/auto/TurnUntilNewTarget.java [hunk 13]
```diff
@@ -23,7 +23,9 @@
 
     private static final Logger logger = Logger.getLogger(TurnUntilNewTarget.class);
     private static final double MAX_ANGULAR_VEL = 5;
-    private static double ANGULAR_VEL = 2.5;
+    private static double FAST_ANGULAR_VEL = 7;
```
Comment (RobinsonZ) [inline src/main/java/org/team1540/robot2019/commands/auto/TurnUntilNewTarget.java]: In a perfect world we'd have a centralized location for all the tuning but I really do not wanna deal with that so let's just jot that down as something to do better next year

File: src/main/java/org/team1540/robot2019/commands/cargo/LoadingStationIntakeCargo.java [hunk 14]
```diff
@@ -4,13 +4,15 @@
...
-import org.team1540.robot2019.commands.hatch.StowHatchMech;
+import org.team1540.robot2019.commands.hatch.simple.ReleaseHatch;
+import org.team1540.robot2019.commands.hatch.simple.RetractHatchMech;
 import org.team1540.robot2019.commands.wrist.WristUp;
 
 public class LoadingStationIntakeCargo extends CommandGroup {
 
     public LoadingStationIntakeCargo() {
-        addSequential(new StowHatchMech());
+        addSequential(new ReleaseHatch());
```
Comment (RobinsonZ) [inline src/main/java/org/team1540/robot2019/commands/cargo/LoadingStationIntakeCargo.java]: Why no more `StowHatchMech`

File: src/main/java/org/team1540/robot2019/commands/cargo/FloorIntakeCargo.java [hunk 15]
```diff
@@ -2,13 +2,15 @@
...
-import org.team1540.robot2019.commands.hatch.StowHatchMech;
+import org.team1540.robot2019.commands.hatch.simple.ReleaseHatch;
+import org.team1540.robot2019.commands.hatch.simple.RetractHatchMech;
 import org.team1540.robot2019.commands.wrist.WristDown;
 
 public class FloorIntakeCargo extends CommandGroup {
 
     public FloorIntakeCargo() {
-        addSequential(new StowHatchMech());
+        addSequential(new ReleaseHatch());
```
Comment (RobinsonZ) [inline src/main/java/org/team1540/robot2019/commands/cargo/FloorIntakeCargo.java]: See below

File: src/main/java/org/team1540/robot2019/commands/drivetrain/deprecated/PercentTankDrive.java [hunk 16]
```diff
@@ -1,4 +1,4 @@
-package org.team1540.robot2019.commands.drivetrain;
+package org.team1540.robot2019.commands.drivetrain.deprecated;
```
Comment (RobinsonZ) [inline src/main/java/org/team1540/robot2019/commands/drivetrain/deprecated/PercentTankDrive.java]: Can we just delete the stuff

File: src/main/java/org/team1540/robot2019/commands/drivetrain/pointdrive/PercentManualLineupLocalizationAngleProvider.java [hunk 17]
```diff
@@ -265,7 +276,7 @@ public double getDistanceToVisionTarget() {
     }
 
     //
-//    private static double smoothStep(double x) {
+//    private static double smoothStep(double x) { // TODO: Method
```
Comment (RobinsonZ) [inline src/main/java/org/team1540/robot2019/commands/drivetrain/pointdrive/PercentManualLineupLocalizationAngleProvider.java]: delet this

File: src/main/java/org/team1540/robot2019/commands/drivetrain/pointdrive/PointManualDriveCommand.java [hunk 18]
```diff
@@ -0,0 +1,156 @@
...
+        } else {
+            logger.warn("Config not set! Setting vel to zero.");
+            twist2DInput.setTwist(Twist2D.ZERO);
+        }
+        pipeline.execute();
+    }
+//
+//    @Override
+//    protected final void usePIDOutput(double output) {
+//
```
Comment (RobinsonZ) [inline src/main/java/org/team1540/robot2019/commands/drivetrain/pointdrive/PointManualDriveCommand.java]: Delet this

File: src/main/java/org/team1540/robot2019/commands/hatch/StowHatchMech.java [hunk 19]
```diff
@@ -1,21 +0,0 @@
-package org.team1540.robot2019.commands.hatch;
```
Comment (RobinsonZ) [inline src/main/java/org/team1540/robot2019/commands/hatch/StowHatchMech.java]: y tho

File: src/main/java/org/team1540/robot2019/commands/hatch/temporary/PlaceHatchInLoadingStation.java [hunk 20]
```diff
@@ -1,9 +1,10 @@
-package org.team1540.robot2019.commands.hatch;
+package org.team1540.robot2019.commands.hatch.temporary;
```
Comment (RobinsonZ) [inline src/main/java/org/team1540/robot2019/commands/hatch/temporary/PlaceHatchInLoadingStation.java]: Evidently these aren't temporary so rename package to reflect that

File: src/main/java/org/team1540/robot2019/utils/ControlUtils.java [hunk 21]
```diff
@@ -21,4 +21,33 @@ public static double simpleDeadzone(double input, double deadzone) {
     public static double allVelocityConstraints(double output, double max, double min, double deadzone) {
         return simpleDeadzone(velocityPosNegConstrain(output, max, min), deadzone);
     }
+
+    /**
+     * https://www.desmos.com/calculator/ybuyhcfzgm
```
Comment (RobinsonZ) [inline src/main/java/org/team1540/robot2019/utils/ControlUtils.java]: A tad unorthodox documentation-wise but uh I'll allow it

File: src/main/java/org/team1540/robot2019/utils/MiniPID.java [hunk 22]
```diff
@@ -0,0 +1,461 @@
+package org.team1540.robot2019.utils;
+
+/**
```
Comment (RobinsonZ) [inline src/main/java/org/team1540/robot2019/utils/MiniPID.java]: Something something plagarism but you gave credit so I guess it's good lol

Would be nice to pull this in as a JitPack dependency but the man didn't check in his hecking build files

**Review by RobinsonZ — COMMENTED** [review]
A good chunk of these are notes to myself

File: src/main/java/org/team1540/robot2019/utils/ControlUtils.java (hunk 21 above)
Comment (wangl5) [reply src/main/java/org/team1540/robot2019/utils/ControlUtils.java]: hey, this is high-quality documentation

**Review by liamjwang — APPROVED** [review]

**Pull request merged**